### ISSUE TITLE
Search failed error when searching from Find Anything

### ISSUE THREAD

**Issue by noziwatele** [netdisco/netdisco-docker#61]
Hi Netdiscoers,

I'm running netdisco-docker and since upgrading to 2.70 (and now 2.70.2) we're getting a Search failed error fairly regularly when searching for devices from the "Find Anything" search function.

I don't see any obvious errors but I don't really know where to look, tbh. If someone could point me in the right direction that would be much appreciated!

## Expected Behavior

Search finds device and displays the Details page.

## Current Behavior

The following error is displayed:
Search failed! Please contact your site administrator (server error).

Often searching again will work without error. Alternately, sometimes switching tabs to Neighbors or Modules and then back to Details will work.

## Steps to Reproduce (for bugs)

As above.
1. Search for a device that is in the DB from Find Anything (either in the menu bar or from the front page)
2. Search fails (server error)

**Comment by rc9000** [comment]
Just to make sure, did you execute netdisco-deploy to upgrade the database schema after the upgrade of the containers? There is a new table in 2.70.

**Comment by noziwatele** [comment]
Thanks for your quick response, @rc9000.

I didn't run netdisco-deploy as I'm using the DB image from docker-hub and I thought that the DB schema updated automatically when the image starts. I saw the following logs on first startup after pulling the new images:

```
netdisco-db-entrypoint: bringing schema up-to-date
netdisco-db-entrypoint: marking schema as up-to-date

```

Is this correct, or is there something else I need to do?

Thanks!

**Comment by ollyg** [comment]
I wonder if there is any clue in `/home/netdisco/logs/netdisco-web.log` ?

**Comment by noziwatele** [comment]
Indeed it does seem like a DB schema issue. Seeing the following error repeatedly in `/home/netdisco/logs/netdisco-web.log`:


```
DBIx::Class::Schema::Versioned::_on_connect(): Your DB is currently unversioned. Please call upgrade on your schema to sync the DB. at /home/netdisco/perl5/lib/perl5/DBICx/Sugar.pm line 121
Error while loading /home/netdisco/perl5/bin/netdisco-web-fg: The setting session_cookie_key must be defined at /home/netdisco/perl5/lib/perl5/Dancer/Session/Cookie.pm line 38.
Compilation failed in require at /home/netdisco/perl5/bin/netdisco-web-fg line 59.
BEGIN failed--compilation aborted at /home/netdisco/perl5/bin/netdisco-web-fg line 59.
```

**Comment by noziwatele** [comment]
And this when the Search Failed error happens:

`[1601] 2023-11-24 09:46:11 error request to GET /ajax/content/device/details crashed: Warning caught during route execution: DBIx::Class::ResultSet::_resolved_attrs(): Useless use of distinct on a grouped resultset ('distinct' is ignored when a 'group_by' is present) at /home/netdisco/perl5/lib/perl5/App/Netdisco/DB/ResultSet/Device.pm line 187`

**Comment by ollyg** [comment]
second error possibly relates to https://github.com/netdisco/netdisco/commit/82ae76e7b7ac09738f05b5b6ace14d2a5217886f
could remove the distinct after we added the group_by

(although need to test if I can reproduce this and it's not a side effect of the first error which is possibly an offline database or other database operational issue)

**Comment by rc9000** [comment]
@noziwatele yes you're of course right, the netdisco variant does the schema upgrade by itself. I've tried now with an old 2.6xx I had lying around, same error after a docker-compose pull. See attached for full log.
[upgrade.txt](https://github.com/netdisco/netdisco-docker/files/13461234/upgrade.txt)

**Comment by ollyg** [comment]
@rc9000 let me know how I can help. I don't really understand the error (FATAL after server starts OK?)

**Comment by rc9000** [comment]
I suspect the `DBIx::Class::Schema::Versioned::_on_connect(): Your DB is currently unversioned.` is just temporary in the upgrade process: postgres restarts, this crashes the web process, but then it comes back fine after the restart.

The `DBIx::Class::ResultSet::_resolved_attrs(): Useless use of distinct`appears after searching, when clicking on the name of a result device and opening eg. `/device?tab=details&q=10.10.1.1`. However the page still opens.

I'm still at a loss how to reproduce a sporadic "search failed", didn't happen for me so far. But I have a practically empty db with only two pseudo devices.

**Comment by rc9000** [comment]
@noziwatele maybe you can describe what exactly you search for, there are various code paths there:

1. ip address or hostname
2. do you select a name from the autocompletion or do you just hit enters
3. is this a fqdn or not, and do you have `domain_suffix`es configured?

**Comment by ollyg** [comment]
Have released Netdisco 2.070003 with fix, should be on container repos within an hour.

**Comment by noziwatele** [comment]
Upgraded to 2.070003 and the issue appears to be resolved! Many thanks for the quick fix!!

Here's more detail on what I was searching for when the issue occurred - in case this information is useful in the future:

1. Was initially searching for hostname, but the error also occurred when I searched for an IP address of a device that was in the DB. But there was no issue when I searched for an IP or MAC address of a Node.
2. I typed the hostname (or partial hostname) and hit enter, which autofilled the FQDN in the search box, then hit enter again and that's when it failed. Usually if I followed the same steps a second time the search would succeed. When it failed on searching for an IP the IP address was autocompleted to the FQDN of the device in the DB, so then I hit enter again and it was the same as for the hostname search.
3. FQDN. Yes we do have domain_suffix configured. Just one suffix that matches what we use on all our devices.

Thanks again!
